packages/headless-react/src/client-wrapper.ts

```diff
@@ -0,0 +1,31 @@
+'use client';
+
+import {useEffect, useReducer, useRef} from 'react';
+
+/**
+ * Alternate for `useSyncExternalStore` which runs into infinite loops when hooks are used in `getSnapshot`
+ * https://github.com/facebook/react/issues/24529
+ */
+export function useSyncMemoizedStore<T>(
```

Comment:
I think we should have unit tests for this.

Some tests I would like to have:
1. When changing the `subscribe` function, unsubscribes then re-subscribes to the new function.
2. When changing `getSnapshot`, should replace the current snapshot by calling `getSnapshot` again.
  * This may not pass currently. One way to make it pass may be to replace `useRef(...)` with `useMemo(() => createRef(...), [getSnapshot])`.
3. Passing the value returned from `useSyncMemoizedStore` to the deps array of a hook (like `useEffect` or `useMemo`) only causes a re-render when the listener of `subscribe` is called.